clcache.py

```diff
@@ -1600,57 +1574,89 @@ def processCompileRequest(cache, compiler, args):
+            return ensureArtifactsExist(cache, artifactSection, cachekey, unusableManifestMissReason,
+                                        objectFile, compilerResult)
+
+        entry = createManifestEntry(manifestHash, includePaths)
+        cachekey = entry.objectHash
 
-                    return getOrSetArtifacts(
-                        cache, cachekey, objectFile, compiler, cmdLine, Statistics.registerEvictedMiss)
-            except IncludeNotFoundException:
-                pass
+        def addManifest():
+            manifest = createOrUpdateManifest(manifestSection, manifestHash, entry)
+            manifestSection.setManifest(manifestHash, manifest)
 
-        return postprocessUnusableManifestMiss(
-            cache, objectFile, manifestSection, manifestHash, sourceFile, compiler, cmdLine,
-            Statistics.registerHeaderChangedMiss)
+        artifactSection = cache.compilerArtifactsRepository.section(cachekey)
+        return ensureArtifactsExist(cache, artifactSection, cachekey, unusableManifestMissReason,
+                                    objectFile, compilerResult, addManifest)
 
 
 def processNoDirect(cache, objectFile, compiler, cmdLine, environment):
     cachekey = CompilerArtifactsRepository.computeKeyNodirect(compiler, cmdLine, environment)
-    return getOrSetArtifacts(cache, cachekey, objectFile, compiler, cmdLine, Statistics.registerCacheMiss, environment)
-
-
-def getOrSetArtifacts(cache, cachekey, objectFile, compiler, cmdLine, statsField, environment=None):
     artifactSection = cache.compilerArtifactsRepository.section(cachekey)
-    cleanupRequired = False
     with artifactSection.lock:
         if artifactSection.hasEntry(cachekey):
             return processCacheHit(cache, objectFile, cachekey)
 
-        compilerResult = invokeRealCompiler(compiler, cmdLine, captureOutput=True, environment=environment)
-        returnCode, compilerStdout, compilerStderr = compilerResult
-        with cache.statistics.lock, cache.statistics as stats:
-            statsField(stats)
-            if returnCode == 0 and os.path.exists(objectFile):
-                artifacts = CompilerArtifacts(objectFile, compilerStdout, compilerStderr)
-                cleanupRequired = addObjectToCache(stats, cache, artifactSection, cachekey, artifacts)
+    compilerResult = invokeRealCompiler(compiler, cmdLine, captureOutput=True, environment=environment)
+
+    return ensureArtifactsExist(cache, artifactSection, cachekey, Statistics.registerCacheMiss,
+                                objectFile, compilerResult)
+
 
-    return compilerResult + (cleanupRequired,)
+def ensureArtifactsExist(cache, section, cachekey, reason, objectFile, compilerResult, extraCallable=None):
+    cleanupRequired = False
+    returnCode, compilerOutput, compilerStderr = compilerResult
+    with section.lock:
+        if not section.hasEntry(cachekey):
+            with cache.statistics.lock, cache.statistics as stats:
+                reason(stats)
```

Comment:
I'm not sure about this - in the old code, the statistics would only get updated with a cache miss if there was no entry in the cache yet. For instance, two concurrent clcache invocations which both produce a cache miss for the same same object would only cause one object to get added to the cache, and the stats only to get bumped once. I.e. the number of added entries would match the number of logged cache misses. This is because clcache would first lock the section and then test whether there's an entry yet.

In that sense, the proposed patch is faithful to the previous behaviour. However, I tend to agree that maybe this is not what we actually want. I'm not sure, it requires more consideration.